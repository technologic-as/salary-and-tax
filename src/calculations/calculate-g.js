import round from './round';

export const calculateG = g => round(99858 * g);

export default calculateG;
