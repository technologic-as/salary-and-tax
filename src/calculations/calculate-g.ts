import round from './round';
import { taxConstants } from '../constants';

export const calculateG = (g: number) => round(taxConstants.oneG * g);

export default calculateG;
