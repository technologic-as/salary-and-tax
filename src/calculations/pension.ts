import { calculateG } from './calculate-g';
import round from './round';
import { PensionProps, PensionStep } from '../constants';

export const pensionConstants: {
  pension: PensionProps;
} = {
  pension: {
    include: true,
    step1: {
      rate: 2,
      threshold: { low: calculateG(1), high: calculateG(7.1) },
    },
    step2: {
      rate: 2,
      threshold: { low: calculateG(7.1), high: calculateG(12) },
    },
  },
};

const calculateStep = (income: number, { rate, threshold: { low, high } }: PensionStep, include: boolean): number => {
  if (!include) {
    return 0;
  }
  const maxStep = high - low;
  const aboveThreshold = income - low;
  const pension = (Math.min(maxStep, aboveThreshold) * rate) / 100;

  return round(aboveThreshold > 0 ? pension : 0);
};

export const calculateStep2 = (salary: number, { step2, include }: PensionProps) => {
  return calculateStep(salary, step2, include);
};

export const calculateStep1 = (salary: number, { step1, include }: PensionProps) => {
  return calculateStep(salary, step1, include);
};

type PensionCalculation = {
  pension: number;
  step1: number;
  step2: number;
  after: number;
};

export const getPension = (salary: number, { pension }: { pension: PensionProps }): PensionCalculation => {
  const step1 = calculateStep1(salary, pension);
  const step2 = calculateStep2(salary, pension);
  const result = pension.include ? round(step1 + step2) : 0;

  return { pension: result, step1, step2, after: salary - result };
};

export default getPension;
