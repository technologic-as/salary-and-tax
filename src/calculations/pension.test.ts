import { calculateG } from './calculate-g';
import { calculateStep1, calculateStep2, getPension, pensionConstants } from './pension';

const tests = [0, 1, 2, 3, 7, 7.1, 8, 11, 12, 15];

describe('pension', () => {
  it('should have defaults matching snapshot', () => {
    expect(pensionConstants).toMatchSnapshot();
  });
  tests.map((g) => {
    const salary = calculateG(g);
    describe(`given salary of ${salary} (${g}G)`, () => {
      it('getPension should match snapshot', () => {
        expect(getPension(salary, pensionConstants)).toMatchSnapshot();
      });
      it('should not calculate if include is false', () => {
        const noPensionData = {
          pension: { ...pensionConstants.pension, include: false },
        };
        expect(getPension(salary, noPensionData)).toMatchSnapshot();
        expect(calculateStep1(salary, noPensionData.pension)).toEqual(0);
        expect(calculateStep2(salary, noPensionData.pension)).toEqual(0);
      });
      it('step 1 should never be greater than 6% of 7.1-1 G', () => {
        expect(calculateStep1(salary, pensionConstants.pension)).not.toBeGreaterThan(35459);
      });
      it('step 2 should never be greater than 6% of 12-7.1 G', () => {
        expect(calculateStep2(salary, pensionConstants.pension)).not.toBeGreaterThan(28483);
      });
      it('getPension should never be greater than 6% of 12-1 G', () => {
        expect(getPension(salary, pensionConstants).pension).not.toBeGreaterThan(63942);
      });
      if (g <= 7.1) {
        it(`step 2 should be 0 when salary is below or equal 7.1 G`, () => {
          expect(calculateStep2(salary, pensionConstants.pension)).toEqual(0);
          expect(getPension(salary, pensionConstants).step2).toEqual(0);
        });
      }
    });
  });
});
