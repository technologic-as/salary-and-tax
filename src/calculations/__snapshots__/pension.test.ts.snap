// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pension given salary of 0 (0G) getPension should match snapshot 1`] = `
Object {
  "after": 0,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 0 (0G) should not calculate if include is false 1`] = `
Object {
  "after": 0,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 111477 (1G) getPension should match snapshot 1`] = `
Object {
  "after": 111477,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 111477 (1G) should not calculate if include is false 1`] = `
Object {
  "after": 111477,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 222954 (2G) getPension should match snapshot 1`] = `
Object {
  "after": 220725,
  "pension": 2229,
  "step1": 2229,
  "step2": 0,
}
`;

exports[`pension given salary of 222954 (2G) should not calculate if include is false 1`] = `
Object {
  "after": 222954,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 334431 (3G) getPension should match snapshot 1`] = `
Object {
  "after": 329972,
  "pension": 4459,
  "step1": 4459,
  "step2": 0,
}
`;

exports[`pension given salary of 334431 (3G) should not calculate if include is false 1`] = `
Object {
  "after": 334431,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 780339 (7G) getPension should match snapshot 1`] = `
Object {
  "after": 766962,
  "pension": 13377,
  "step1": 13377,
  "step2": 0,
}
`;

exports[`pension given salary of 780339 (7G) should not calculate if include is false 1`] = `
Object {
  "after": 780339,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 791486 (7.1G) getPension should match snapshot 1`] = `
Object {
  "after": 777886,
  "pension": 13600,
  "step1": 13600,
  "step2": 0,
}
`;

exports[`pension given salary of 791486 (7.1G) should not calculate if include is false 1`] = `
Object {
  "after": 791486,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 891816 (8G) getPension should match snapshot 1`] = `
Object {
  "after": 876210,
  "pension": 15606,
  "step1": 13600,
  "step2": 2006,
}
`;

exports[`pension given salary of 891816 (8G) should not calculate if include is false 1`] = `
Object {
  "after": 891816,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 1226247 (11G) getPension should match snapshot 1`] = `
Object {
  "after": 1203952,
  "pension": 22295,
  "step1": 13600,
  "step2": 8695,
}
`;

exports[`pension given salary of 1226247 (11G) should not calculate if include is false 1`] = `
Object {
  "after": 1226247,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 1337724 (12G) getPension should match snapshot 1`] = `
Object {
  "after": 1313200,
  "pension": 24524,
  "step1": 13600,
  "step2": 10924,
}
`;

exports[`pension given salary of 1337724 (12G) should not calculate if include is false 1`] = `
Object {
  "after": 1337724,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension given salary of 1672155 (15G) getPension should match snapshot 1`] = `
Object {
  "after": 1647631,
  "pension": 24524,
  "step1": 13600,
  "step2": 10924,
}
`;

exports[`pension given salary of 1672155 (15G) should not calculate if include is false 1`] = `
Object {
  "after": 1672155,
  "pension": 0,
  "step1": 0,
  "step2": 0,
}
`;

exports[`pension should have defaults matching snapshot 1`] = `
Object {
  "pension": Object {
    "include": true,
    "step1": Object {
      "rate": 2,
      "threshold": Object {
        "high": 791486,
        "low": 111477,
      },
    },
    "step2": Object {
      "rate": 2,
      "threshold": Object {
        "high": 1337724,
        "low": 791486,
      },
    },
  },
}
`;
